### PR TITLE
Use ruff as a linter

### PR DESCRIPTION
**Pull request by vshampor**
### Changes
Add ruff as a step to the `pre-commit` pipeline and adjust the codebase to conform.

### Reason for changes
Ruff seems to do a good job at basic linting, works much faster than pylint and doesn't even seem to need a virtual environment to function properly. We might want to replace pylint with this entirely to lessen the load on CI.

### Related tickets
N/A

### Tests
precommit

### CODE REVIEW

File: examples/torch/common/model_loader.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ def is_url(uri):
     try:
         parsed_url = urllib.parse.urlparse(uri)
         return parsed_url.scheme and parsed_url.netloc
-    except:  # pylint: disable=bare-except
+    except:  # pylint: disable=bare-except  # noqa: E722
```
Comment (AlexanderDokuchaev) [inline examples/torch/common/model_loader.py:95]: Is it possible to use human readable rule name, like `bare-except`?

Comment (vshampor) [reply examples/torch/common/model_loader.py:95]: Sadly, no.

**Comment by AlexanderDokuchaev** [comment]
Can you add check for copyrights like https://github.com/openvinotoolkit/nncf/pull/1833?
Ruff uses flake8-copyright to it by set `copyright` in config.

**Comment by vshampor** [comment]
> Can you add check for copyrights like #1833? Ruff uses flake8-copyright to it by set `copyright` in config.

Done

File: nncf/torch/pruning/filter_pruning/global_ranking/evolutionary_optimization.py [hunk 2]
```diff
@@ -270,7 +270,7 @@ class LeGRPruner:
     and resetting all changes in the model made by the environment.
     """
 
-    def __init__(self, filter_pruner_ctrl: "FilterPruningController", model: nn.Module):
+    def __init__(self, filter_pruner_ctrl: "FilterPruningController", model: nn.Module):  # noqa: F821
```
Comment (KodiaqQ) [inline nncf/torch/pruning/filter_pruning/global_ranking/evolutionary_optimization.py:273]: I'm afraid we'll have to insert multiple `noqa`'s into the section where the `pylint` agreed with the code.

File: nncf/__init__.py [hunk 3]
```diff
@@ -12,28 +12,28 @@
...
-from nncf.parameters import ModelType
-from nncf.parameters import TargetDevice
-from nncf.quantization import QuantizationPreset
-from nncf.quantization import compress_weights
-from nncf.quantization import quantize
-from nncf.quantization import quantize_with_accuracy_control
-from nncf.quantization.advanced_parameters import AdvancedQuantizationParameters
-from nncf.scopes import IgnoredScope
-from nncf.version import __version__
+from nncf.common.logging import nncf_logger as nncf_logger
```
[the remark this answers is not in the document]
Comment (vshampor) [reply nncf/__init__.py:15]: This is one of the flake8 or mypy rules integrated into ruff by default to enforce the policy of clearly defining just what imports in the `__init__.py` files are to be considered "reexports", e.g. to provide aliases for API, for instance. 

https://stackoverflow.com/questions/72504576/why-use-from-module-import-a-as-a-instead-of-just-from-module-import-a

Perhaps we could use `__all__` instead, or disable this rule somehow via ruff config, but I actually find it good to have some extra visual cues to symbols that are meant to be aliases, or API, and the fact that one has to do some extra work and therefore be extra sure that they want to do aliasing and not just add an unused import.

File: nncf/common/utils/logger/__init__.py [hunk 4]
```diff
@@ -20,6 +20,6 @@
 )
 
 # pylint:disable=unused-import
```
[the remark this answers is not in the document]
Comment (vshampor) [reply nncf/common/utils/logger/__init__.py:22]: We will, eventually, first step is to get ruff working and then, if it works fine, we'll drop pylint

File: nncf/experimental/torch/nas/bootstrapNAS/__init__.py [hunk 5]
```diff
@@ -8,6 +8,8 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+# noqa
```
[the remark this answers is not in the document]
Comment (vshampor) [reply nncf/experimental/torch/nas/bootstrapNAS/__init__.py:12]: Right, @ljaljushkin might want to take a look at that later

File: nncf/experimental/torch/nas/bootstrapNAS/search/search.py [hunk 6]
```diff
@@ -164,7 +163,7 @@ def run(
         checkpoint_save_dir: str,
         efficiency_evaluator: Optional[BaseEvaluator] = None,
         ref_acc: Optional[float] = 100,
-        tensorboard_writer: Optional[SummaryWriter] = None,
+        tensorboard_writer: Optional["SummaryWriter"] = None,  # noqa: F821
```
[the remark this answers is not in the document]
Comment (vshampor) [reply nncf/experimental/torch/nas/bootstrapNAS/search/search.py:166]: The `SummaryWriter` was imported merely for purposes for type hinting. At the same time the import brings an implicit dependency on a `tensorboard` package being installed, but we don't have that as hard dependency.

File: nncf/torch/pruning/filter_pruning/global_ranking/evolutionary_optimization.py (hunk 2 above)
Comment (vshampor) [reply nncf/torch/pruning/filter_pruning/global_ranking/evolutionary_optimization.py:273]: Only in places where we could not import a corresponding class to have a type hint due to circular dependencies.

**Review by KodiaqQ — APPROVED** [review]

**Review by AlexanderDokuchaev — APPROVED** [review]

**Pull request merged**